### DISCUSSION TITLE
Time field

### DISCUSSION THREAD

**Discussion by manson** [kamalshkeir/kago#14] [General]
Hi!
New small issue in ORM
I write this note as I remember I faced, because when struggling with it, I have rewritten code many times and may be wrong in some nuances, but all in all something like that exists:
I have a time field, say, LockStart. My local time is not a UTC time. When I save this field in model like time.Now() it is converted to UTC and saves in database as UTC. It's ok for me, although someone would argue that he sees the time in the table, not he wrote.  But to be clear how it works and stores, I insert model with this field like 
```LockStart:    time.Now().UTC(),```. It's ok. it shifts time to a UTC time and is ok. But, when I update record and set this file like ```.Set("lock_start", time.Now().UTC())``` it behaves weird. The time is intact (not shifted to a valid UTC) but extra info, when printing this field to console, says that this is UTC time, although it is already local.
In other words (I skip here dates to accent the problem part) when I insert record as time.Now().UTC() (my local time 17:00 is being converted to 14:00 UTC) it shows 14:00 UTC as expected. But when I updated it the same way (Set field with  time.Now().UTC()), when I read this field back I gets 17:00 UTC. But I expected 14:00 UTC. 

I hope I didn't mess up anything here. Currently I have worked around this like having this field as int64 and saving Time as Unix value. It is even better for me like that in this situation, but anyway, something is wrong with this field type when saving, updating and reading. I know, time fields always was pain.

**Comment by kamalshkeir** [comment]
Hi @manson ,
I couldn't reproduce the same behavior, are you using last version of korm ? , here is what i tried : 
```go
package main

import (
	"fmt"
	"time"

	"github.com/kamalshkeir/klog"
	"github.com/kamalshkeir/korm"
	"github.com/kamalshkeir/sqlitedriver"
)

type TestModel struct {
	Id        uint `korm:"pk"`
	Content   string
	TimeField time.Time
}

func main() {
	sqlitedriver.Use()
	err := korm.New(korm.SQLITE, "db")
	if klog.CheckError(err) {
		return
	}
	defer korm.Shutdown()

	err = korm.AutoMigrate[TestModel]("test_model")
	if klog.CheckError(err) {
		return
	}
	klog.Printfs("mgrunning on http://localhost:9313\n")

	m := TestModel{
		Content:   "test3",
		TimeField: time.Now().UTC(),
	}
	fmt.Println("before insert:", m.TimeField)
	_, err = korm.Model[TestModel]().Insert(&m)
	klog.CheckError(err)
	time.Sleep(5 * time.Second)
	//update
	_, err = korm.Model[TestModel]().Where("content = ?", m.Content).Set("time_field = ?", time.Now().UTC())
	klog.CheckError(err)
	// after update
	data, err := korm.Model[TestModel]().Where("content = ?", m.Content).One()
	if klog.CheckError(err) {
		return
	}
	fmt.Println("after update:", data.TimeField)
}
```

OUTPUT:
```sh
before insert: 2023-02-05 15:55:49.8155274 +0000 UTC
after update: 2023-02-05 15:55:54 +0000 UTC
```

**Reply by manson** [reply]
Oh, I tried so many times to fix this that forgot the sequence. I've checked your example and indeed it works. Although with UTC it works now (It wasn't before) but I see the same issue in this example if you remove UTC()  in insert as well in update. Just leave time.Now() only. I get this:
```
before insert: 2023-02-05 22:40:50.307710959 +0300 MSK m=+0.031145443
after update: 2023-02-05 22:40:55 +0000 UTC
```
First is local, second UTC but in fact is local too...

**Reply by manson** [reply]
A bit more complex example. I've added read value from db after insert:
```
func main() {
	sqlitedriver.Use()
	err := korm.New(korm.SQLITE, "db")
	if klog.CheckError(err) {
		return
	}
	defer korm.Shutdown()

	err = korm.AutoMigrate[TestModel]("test_model")
	if klog.CheckError(err) {
		return
	}
	klog.Printfs("mgrunning on http://localhost:9313\n")

	m := TestModel{
		Content:   "test3",
		TimeField: time.Now(),
	}
	fmt.Println("before insert:", m.TimeField)
	_, err = korm.Model[TestModel]().Insert(&m)
	klog.CheckError(err)
	data, err := korm.Model[TestModel]().Where("content = ?", m.Content).One()
	if klog.CheckError(err) {
		return
	}
	fmt.Println("after insert:", data.TimeField)
	time.Sleep(5 * time.Second)
	//update
	_, err = korm.Model[TestModel]().Where("content = ?", m.Content).Set("time_field = ?", time.Now())
	klog.CheckError(err)
	// after update
	data, err = korm.Model[TestModel]().Where("content = ?", m.Content).One()
	if klog.CheckError(err) {
		return
	}
	fmt.Println("after update:", data.TimeField)
}
```
And gets:
```
before insert: 2023-02-05 22:53:18.828057949 +0300 MSK m=+0.007035447
after insert: 2023-02-05 19:52:49 +0000 UTC
after update: 2023-02-05 22:53:23 +0000 UTC

```

**Reply by kamalshkeir** [reply]
@manson worth noting that sqlite doesn't have date time column type, only string , so what i have tried to do, is that make it work with all dialects when sorting and filtering even if string in sqlite, and also with date time format coming from html and js , so you can sometime found a time.Time column with some extra suffix saved as string

**Reply by manson** [reply]
> @manson worth noting that sqlite doesn't have date time column type, only string , so what i have tried to do, is that make it work with all dialects when sorting and filtering even if string in sqlite, and also with date time format coming from html and js , so you can sometime found a time.Time column with some extra suffix saved as string

I understand. But extra suffixes is not a problem in most cases - all convertions from string back to time usually understans this.  But what if to save data in unix type (unixnano)? It is a int64 value, besides well ordering. Dont know how you order datetime field now by sql queries... And it doesn't care about time zones - what value you save - that you gets back and restoring datetime is solely you problem (you know what is saved there). 
But if you want to keep datetime fields for some databases they support, you may want to create your own datetime type for KORM in Go that would keep datetime in unix format in all dialects (all support int64) in database, and if user wants specific databases datetime type of that dialect, let him use it as it is now. I think this is a good decision. 
Example:

```CreatedAt    KormTime 'korm:"now"'```

In this case before saving the datetime value always normalizes so no matter in what form it came, finally it will have the universal understandable format.

**Reply by kamalshkeir** [reply]
so why not doing it like this : 
```go
package main

import (
	"fmt"
	"time"

	"github.com/kamalshkeir/klog"
	"github.com/kamalshkeir/korm"
	"github.com/kamalshkeir/sqlitedriver"
)

type TestModel struct {
	Id        uint `korm:"pk"`
	Content   string
	TimeUnix  int64
	TimeField time.Time
}

func main() {
	sqlitedriver.Use()
	err := korm.New(korm.SQLITE, "db")
	if klog.CheckError(err) {
		return
	}
	defer korm.Shutdown()

	err = korm.AutoMigrate[TestModel]("test_model")
	if klog.CheckError(err) {
		return
	}

	m := TestModel{
		Content:   "test4",
		TimeUnix:  time.Now().Unix(),
		TimeField: time.Now(),
	}
	fmt.Println("before insert:", m.TimeUnix)
	fmt.Println("before insert:", m.TimeField)
	_, err = korm.Model[TestModel]().Insert(&m)
	klog.CheckError(err)
	data, err := korm.Model[TestModel]().Where("content = ?", m.Content).One()
	if klog.CheckError(err) {
		return
	}
	fmt.Println("after insert:", data.TimeUnix)
	fmt.Println("after insert:", data.TimeField)
	time.Sleep(5 * time.Second)
	//update
	_, err = korm.Model[TestModel]().Where("content = ?", m.Content).Set("time_field = ?,time_unix = ?", time.Now(), time.Now().Unix())
	klog.CheckError(err)
	// after update
	data, err = korm.Model[TestModel]().Where("content = ?", m.Content).One()
	if klog.CheckError(err) {
		return
	}
	fmt.Println("after update:", time.Unix(data.TimeUnix, 0))
	fmt.Println("after update:", data.TimeField)
}

```

**Reply by kamalshkeir** [reply]
this is the output of the same code with Postgres driver:

```sh
before insert: 1675634320
before insert: 2023-02-05 22:58:40.7351038 +0100 CET m=+0.682909301
after insert: 1675634320
after insert: 2023-02-05 22:58:40 +0000 +0000
after update: 2023-02-05 22:58:45 +0100 CET
after update: 2023-02-05 22:58:45.824202 +0000 +0000
```

With sqlite:
```sh
before insert: 1675634379
before insert: 2023-02-05 22:59:39.5285877 +0100 CET m=+0.066513501
after insert: 1675634379
after insert: 2023-02-05 22:59:39 +0000 UTC
after update: 2023-02-05 22:59:44 +0100 CET
after update: 2023-02-05 22:59:44 +0000 UTC
```

With mysql:
```sh
before insert: 1675634674
before insert: 2023-02-05 23:04:34.915804 +0100 CET m=+0.763712001
after insert: 1675634674
after insert: 2023-02-05 23:04:34 +0000 UTC
after update: 2023-02-05 23:04:40 +0100 CET
after update: 2023-02-05 22:04:40 +0000 UTC
```

**Reply by kamalshkeir** [reply]
It's soooo hard @manson  ! 😓, here is what i was comparing, seems the same for me, no ? :

```go
package main

import (
	"fmt"
	"time"

	"github.com/kamalshkeir/klog"
	"github.com/kamalshkeir/korm"
	"github.com/kamalshkeir/sqlitedriver"
)

type TestModel struct {
	Id        uint `korm:"pk"`
	Content   string
	TimeUnix  int64
	TimeField time.Time
}

func main() {
	sqlitedriver.Use()
	err := korm.New(korm.SQLITE, "db")
	if klog.CheckError(err) {
		return
	}
	defer korm.Shutdown()

	err = korm.AutoMigrate[TestModel]("test_model")
	if klog.CheckError(err) {
		return
	}

	t := time.Now()
	m := TestModel{
		Content:   "new1",
		TimeUnix:  t.UTC().Unix(),
		TimeField: t.UTC(),
	}
	fmt.Println("before insert unix:", time.Unix(m.TimeUnix, 0), ", utc: ", time.Unix(m.TimeUnix, 0).UTC())
	fmt.Println("before insert:", m.TimeField, ", utc: ", m.TimeField.UTC())
	_, err = korm.Model[TestModel]().Insert(&m)
	klog.CheckError(err)
	data, err := korm.Model[TestModel]().Where("content = ?", m.Content).One()
	if klog.CheckError(err) {
		return
	}
	fmt.Println("after insert unix:", time.Unix(data.TimeUnix, 0), ", utc: ", time.Unix(data.TimeUnix, 0).UTC())
	fmt.Println("after insert:", data.TimeField, ", utc: ", data.TimeField.UTC())
	time.Sleep(5 * time.Second)
	//update
	_, err = korm.Model[TestModel]().Where("content = ?", m.Content).Set("time_field = ?,time_unix = ?", time.Now().UTC(), time.Now().UTC().Unix())
	klog.CheckError(err)
	// after update
	data, err = korm.Model[TestModel]().Where("content = ?", m.Content).One()
	if klog.CheckError(err) {
		return
	}
	fmt.Println("after update unix:", time.Unix(data.TimeUnix, 0), ", utc: ", time.Unix(data.TimeUnix, 0).UTC())
	fmt.Println("after update:", data.TimeField, ", utc: ", data.TimeField.UTC())
}

```

OUTPUT:
```sh
before insert unix: 2023-02-06 00:22:41 +0100 CET ,utc:  2023-02-05 23:22:41 +0000 UTC
before insert: 2023-02-05 23:22:41.3787759 +0000 UTC , utc:  2023-02-05 23:22:41.3787759 +0000 UTC

after insert unix: 2023-02-06 00:22:41 +0100 CET , utc:  2023-02-05 23:22:41 +0000 UTC
after insert: 2023-02-05 23:22:41 +0000 UTC , utc:  2023-02-05 23:22:41 +0000 UTC

after update unix: 2023-02-06 00:22:46 +0100 CET , utc:  2023-02-05 23:22:46 +0000 UTC
after update: 2023-02-05 23:22:46 +0000 UTC , utc:  2023-02-05 23:22:46 +0000 UTC
```

**Reply by manson** [reply]
I don't see the conclusions you have drawn. As far as I see Unix and UTC works well in all cases except local time.
```
after update: 2023-02-05 22:59:44 +0100 CET
after update: 2023-02-05 22:59:44 +0000 UTC
```
This is the different time, right? So updating by time.Now() works weird.
```so why not doing it like this``` - yes, currently in critical parts where datetime matters, I would use unix time with int64 field type. In other cases (CreatedAt, UpdatedAt) a can leave it as is.
But I would prefer to use the custom Korm type KormTime that is the same as go Time with all it's internal functions, but saves as int64 unix in database. Or even without custom type, just save data in unix format. If user will use Korm for data querying, it will get datetime field after loading. If some other tool - just int64 and this is not the problem if user knows what type stores here. After all, if a user wants to save displayable datetime he may store it as a string as you do it right now. But to avoid possible collisions, it is better to have our own datetime type

**Reply by kamalshkeir** [reply]
@manson , i changed to use INTEGER instead of string for sqlite by default (on local, i didn't push yet), so for this code : 
```go
package main

import (
	"fmt"
	"time"

	"github.com/kamalshkeir/klog"
	"github.com/kamalshkeir/korm"
	"github.com/kamalshkeir/sqlitedriver"
)

type TestModel struct {
	Id        uint `korm:"pk"`
	Content   string
	TimeField time.Time
}

func main() {
	sqlitedriver.Use()
	err := korm.New(korm.SQLITE, "db")
	if klog.CheckError(err) {
		return
	}
	defer korm.Shutdown()

	err = korm.AutoMigrate[TestModel]("test_model")
	if klog.CheckError(err) {
		return
	}
	klog.Printfs("mgrunning on http://localhost:9313\n")

	m := TestModel{
		Content:   "test3",
		TimeField: time.Now(),
	}
	fmt.Println("before insert:", m.TimeField)
	_, err = korm.Model[TestModel]().Insert(&m)
	klog.CheckError(err)
	data, err := korm.Model[TestModel]().Where("content = ?", m.Content).One()
	if klog.CheckError(err) {
		return
	}
	fmt.Println("after insert:", data.TimeField)
	time.Sleep(5 * time.Second)
	//update
	_, err = korm.Model[TestModel]().Where("content = ?", m.Content).Set("time_field = ?", time.Now().UTC())
	klog.CheckError(err)
	// after update
	data, err = korm.Model[TestModel]().Where("content = ?", m.Content).One()
	if klog.CheckError(err) {
		return
	}
	fmt.Println("after update:", data.TimeField)
}
```

i got Output for sqlite:
```sh
before insert: 2023-02-06 10:21:30.1951215 +0100 CET m=+0.018690601
after insert: 2023-02-06 10:21:30 +0000 UTC
after update: 2023-02-06 09:21:35 +0000 UTC
```

Postgres, will be saved in database as timestamptz so with time zone support
```sh
before insert: 2023-02-06 10:33:38.2612184 +0100 CET m=+0.196264201
after insert: 2023-02-06 09:33:38.261218 +0000 GMT
after update: 2023-02-06 09:33:43.306853 +0000 GMT
```



Better  ?

**Reply by manson** [reply]
For postgress looks ok, no questions but I dont understand the result for sqlite. Results should be like for postgress. Look:
Initial value (before insert) is ```10:21:30 ... CET```. CET - Central European Time. Your local time. It is +1 from UTC. When you saved it and later read you should get ether the same  ```10:21:30  CET``` or ```09:21:30 UTC``` or ```09:21:30 GMT```. I.e. CET is 1 hour ahead of UTC and GMT. And we see this picture for postgress but not for sqlite. For some reason ```after insert``` shows the CET's hour but has UTC in time description. In your case you should see like:
```
before insert: 2023-02-06 10:21:30.1951215 +0100 CET m=+0.018690601
after insert: 2023-02-06 09:21:30 +0000 UTC
after update: 2023-02-06 09:21:35 +0000 UTC
```

In Go this is simple. To save you should do TimeField.Unix() (or UnixNano for better accuracy) and when read, TimeField = time.Unix(fromDbValue, 0) (or time.UnixNano(fromDbValue, 0)). Unix value does know nothing about UTC, GMT or any other local times. It just correctly calculates seconds form 1970 for any of them. So TimeField.Unix() will be equal to TimeField.UTC().Unix() or whatever.  After all, working with Unix time you may forget all other convertions except at final stage when you decide how to show it. I did this convertions manually saving to db and vice versa ant gets the correct values. Probably you do some other conversions if you have invalid ```after insert``` and thats why the result after changing to int64 did not change

Try this simple script and you will get it:
```
t := time.Now().Unix();
t1 := time.Now().UTC().Unix();
// t = t1 

fmt.Println(t) 
fmt.Println(t1) 

fmt.Println(time.Unix(t,0))
fmt.Println(time.Unix(t1,0))

fmt.Println(time.Unix(t,0).UTC())
fmt.Println(time.Unix(t1,0).UTC())

```
you will get something like:
```
1675692542
1675692542
2023-02-06 17:09:02 +0300 MSK
2023-02-06 17:09:02 +0300 MSK
2023-02-06 14:09:02 +0000 UTC
2023-02-06 14:09:02 +0000 UTC
```

So after insert or after update whatever you set in TimeField (time.Now().Unix()  or  time.Now().UTC().Unix() ) the result should be the same

**Reply by kamalshkeir** [reply]
@manson yep you are right, i was changing everything to work with as unix int64, it is working for mysql, i will test other dialects, but should work, so now all time fields are migrated as BIGINT for sql , and can be loaded directly to time.Time when working with builder, the admin dashboard also changed to send and recv unix time instead of string

**Reply by manson** [reply]
@kamalshkeir Great! Just don't forget in dashboard to display datetime fields in readable format as it is now (and dropdown calendar in input field as it is now). Actually, I don't know how is better - in local format or in universal UTC. Both ways have cons and pros.

**Reply by kamalshkeir** [reply]
@manson , working fine, tested for all dialects, You can update to v1.5.5 and make sure you remove assets folder to generate a new one, also you should add korm.WithShell() after WithDashboard when used with admin, and after New if not used with dashboard. it's not automatic anymore , and delete the database also for new migrations

**Reply by kamalshkeir** [reply]
@manson i think it still better like this, the same way for all dialects, tell me what do you think when you check it

**Reply by manson** [reply]
Great, will check it tonight, but I didn't get this:
``` you should add korm.WithShell() after WithDashboard when used with admin, and after New if not used with dashboard.```
Do you mean insterd of ```sbus := korm.WithDashboard()``` I should use 
```
sbus := korm.WithDashboard()
korm.WithShell()
```
...to make shell work?

```delete the database also for new migrations``` - should I do this only once now to use this changes or all the times I change structure?

**Reply by kamalshkeir** [reply]
> Great, will check it tonight, but I didn't get this:
> ``` you should add korm.WithShell() after WithDashboard when used with admin, and after New if not used with dashboard.```
> Do you mean insterd of ```sbus := korm.WithDashboard()``` I should use 
> ```
> sbus := korm.WithDashboard()
> korm.WithShell()
> ```
> ...to make shell work?

yep exactly like this

**Reply by kamalshkeir** [reply]
@manson https://www.youtube.com/watch?v=UDxQHBTHE9g , this is how I like to start my project with korm, unfortunately with no sound because of my bad accent :) , so you can watch it at 2x speed easy, hope you like it

**Reply by manson** [reply]
> @manson https://www.youtube.com/watch?v=UDxQHBTHE9g , this is how I like to start my project with korm, unfortunately with no sound because of my bad accent :) , so you can watch it at 2x speed easy, hope you like it

Watched this. Very well and usefull video. Looks like it is good format for understanding at least how to use this framework along with documents. If you like to continue this tutorials, you whould need to split all topics to parts and from the ground up implement every "online". Say, 
"Environment settings .env",
"Simple API", 
"Websockets server", 
"Using SSE", 
"Use several different databases at once", 
"API request templates and parameters",  
"BUS for local usage", 
"BUS for inter applications communications", 
"ORM usage", 
"ORM and BUS to work together - bind two or more ORM applications", 
"ORM Extra usage (triggers)"
"Logging"
"HTML Templates"
"Admin dashboard"
"Shell"
and so on

If you don't want to speak, it would be nice to on some light musing in the background. So, I feel this positive :-)

**Reply by manson** [reply]
@kamalshkeir 
Just tried new version of ORM. Not all checked yet, but found small issue - why not to show datetime in admin dashbord like datetime, not number?
![image](https://user-images.githubusercontent.com/176285/217752605-b4ef0029-22a1-4708-affd-e4a526ca3623.png)
You know the type, just convert it to a string in UI or during server-side rendering (I mentioned this before)

**Reply by kamalshkeir** [reply]
> @kamalshkeir 
> Just tried new version of ORM. Not all checked yet, but found small issue - why not to show datetime in admin dashbord like datetime, not number?
> ![image](https://user-images.githubusercontent.com/176285/217752605-b4ef0029-22a1-4708-affd-e4a526ca3623.png)
> You know the type, just convert it to a string in UI or during server-side rendering (I mentioned this before)

it is in readable format when you go to single row page, i keep it like this in all rows page because of the search if you are searchinh for specific timestamp, and also make it clear that time is unix

**Reply by kamalshkeir** [reply]
i will try it if it's working correctly in all models page and add it if yes

**Reply by kamalshkeir** [reply]
@manson not working as it should in all models page search when i am trying something like `created_at='2023-02-09T11:36'`, so fo now i will keep it as unix in this page, and when you go to single model by clicking on the id you will have it in a js format, also when creating you will have an input datetime-local

**Reply by manson** [reply]
> @manson not working as it should in all models page search when i am trying something like `created_at='2023-02-09T11:36'`, so fo now i will keep it as unix in this page, and when you go to single model by clicking on the id you will have it in a js format, also when creating you will have an input datetime-local

Ok. Then another issue (or I don't understand something): When adding record by using integrated admin panel, if custom field ends with ```_uuid``` it is locked for entering. And in edit mode  - all the field with ```_uuid``` and ```_id``` are locked. I have my custom fields like ```something_uuid``` and so cannot fill this. I understand why you did this, but I think it is better to permit all fields to edit except ```id``` - if user enters wrong value, it is his responsibility, than to not to permit user input data at all...

**Reply by kamalshkeir** [reply]
@manson but the field something_uuid is auto generated ?
> but I think it is better to permit all fields to edit except id - if user enters wrong value, it is his responsibility, than to not to permit user input data at all...
yeah i agree , i will change it

**Reply by kamalshkeir** [reply]
@manson , done, now only pk field is not editable, and only pk and password not allowed to update 

v1.5.6 and update assets folder

**Reply by manson** [reply]
> @manson but the field something_uuid is auto generated ?

No, I do it manually.

**Reply by manson** [reply]
> @manson , done, now only pk field is not editable, and only pk and password not allowed to update
> 
> v1.5.6 and update assets folder

Super! :-) Thanks
